include/deal.II/base/tensor_base.h

```diff
@@ -1098,276 +1127,157 @@ Tensor<1,dim,Number> &Tensor<1,dim,Number>::operator /= (const Number s)
 }
 
 
-
-template <int dim, typename Number>
-inline
-Number
-Tensor<1,dim,Number>::operator * (const Tensor<1,dim,Number> &p) const
-{
-  // unroll by hand since this is a
-  // frequently called function and
-  // some compilers don't want to
-  // always unroll the loop in the
-  // general template
-  switch (dim)
-    {
-    case 1:
-      return (values[0] * p.values[0]);
-      break;
-    case 2:
-      return (values[0] * p.values[0] +
-              values[1] * p.values[1]);
-      break;
-    case 3:
-      return (values[0] * p.values[0] +
-              values[1] * p.values[1] +
-              values[2] * p.values[2]);
-      break;
-    default:
-      Number q = values[0] * p.values[0];
-      for (unsigned int i=1; i<dim; ++i)
-        q += values[i] * p.values[i];
-      return q;
-    }
-}
-
-
-
-template <int dim, typename Number>
+template <int rank_, int dim, typename Number>
 inline
-Tensor<1,dim,Number> Tensor<1,dim,Number>::operator + (const Tensor<1,dim,Number> &p) const
+Tensor<rank_,dim,Number>
+Tensor<rank_,dim,Number>::operator - () const
 {
-  return (Tensor<1,dim,Number>(*this) += p);
-}
+  Tensor<rank_,dim,Number> tmp;
 
-
-
-template <int dim, typename Number>
-inline
-Tensor<1,dim,Number> Tensor<1,dim,Number>::operator - (const Tensor<1,dim,Number> &p) const
-{
-  return (Tensor<1,dim,Number>(*this) -= p);
-}
-
-
-
-template <int dim, typename Number>
-inline
-Tensor<1,dim,Number> Tensor<1,dim,Number>::operator - () const
-{
-  Tensor<1,dim,Number> result (false);
   for (unsigned int i=0; i<dim; ++i)
-    result.values[i] = -values[i];
-  return result;
-}
+    tmp.values[i] = -values[i];
 
+  return tmp;
+}
 
 
-template <int dim, typename Number>
+template <int rank_, int dim, typename Number>
 inline
-typename Tensor<1,dim,Number>::real_type
-Tensor<1,dim,Number>::norm () const
+typename Tensor<rank_,dim,Number>::real_type
+Tensor<rank_,dim,Number>::norm () const
 {
   return std::sqrt (norm_square());
 }
 
 
-
-template <int dim, typename Number>
+template <int rank_, int dim, typename Number>
 inline
-typename Tensor<1,dim,Number>::real_type
-Tensor<1,dim,Number>::norm_square () const
+typename Tensor<rank_,dim,Number>::real_type
+Tensor<rank_,dim,Number>::norm_square () const
 {
-  real_type s = numbers::NumberTraits<Number>::abs_square(values[0]);
-  for (unsigned int i=1; i<dim; ++i)
-    s += numbers::NumberTraits<Number>::abs_square(values[i]);
+  real_type s = 0;
+  for (unsigned int i=0; i<dim; ++i)
+    s += static_cast<Tensor<rank_-1,dim,Number> >(values[i]).norm_square();
```

Comment:
As below.

I would be more than happy to remove the special notion of ```Tensor<1,dim>```s value type.